vnext/Shared/JSI/NapiJsiRuntime.cpp

```diff
@@ -0,0 +1,1921 @@
+  EnvScope scope{m_env};
+  m_propertyId.Error = NapiRefHolder{this, GetPropertyIdFromName("Error"sv)};
+  m_propertyId.Object = NapiRefHolder{this, GetPropertyIdFromName("Object"sv)};
+  m_propertyId.Proxy = NapiRefHolder{this, GetPropertyIdFromName("Proxy"sv)};
+  m_propertyId.Symbol = NapiRefHolder{this, GetPropertyIdFromName("Symbol"sv)};
+  m_propertyId.byteLength = NapiRefHolder{this, GetPropertyIdFromName("byteLength"sv)};
+  m_propertyId.configurable = NapiRefHolder{this, GetPropertyIdFromName("configurable"sv)};
+  m_propertyId.enumerable = NapiRefHolder{this, GetPropertyIdFromName("enumerable"sv)};
+  m_propertyId.get = NapiRefHolder{this, GetPropertyIdFromName("get"sv)};
+  m_propertyId.getOwnPropertyDescriptor = NapiRefHolder{this, GetPropertyIdFromName("getOwnPropertyDescriptor"sv)};
+  m_propertyId.hostFunctionSymbol = NapiRefHolder{this, CreateSymbol("hostFunctionSymbol"sv)};
+  m_propertyId.hostObjectSymbol = NapiRefHolder{this, CreateSymbol("hostObjectSymbol"sv)};
+  m_propertyId.length = NapiRefHolder{this, GetPropertyIdFromName("length"sv)};
+  m_propertyId.message = NapiRefHolder{this, GetPropertyIdFromName("message"sv)};
+  m_propertyId.ownKeys = NapiRefHolder{this, GetPropertyIdFromName("ownKeys"sv)};
+  m_propertyId.propertyIsEnumerable = NapiRefHolder{this, GetPropertyIdFromName("propertyIsEnumerable"sv)};
+  m_propertyId.prototype = NapiRefHolder{this, GetPropertyIdFromName("prototype"sv)};
+  m_propertyId.set = NapiRefHolder{this, GetPropertyIdFromName("set"sv)};
+  m_propertyId.toString = NapiRefHolder{this, GetPropertyIdFromName("toString"sv)};
+  m_propertyId.value = NapiRefHolder{this, GetPropertyIdFromName("value"sv)};
+  m_propertyId.writable = NapiRefHolder{this, GetPropertyIdFromName("writable"sv)};
+
+  m_value.Undefined = NapiRefHolder{this, GetUndefined()};
+  m_value.Null = NapiRefHolder{this, GetNull()};
+  m_value.True = NapiRefHolder{this, GetBoolean(true)};
+  m_value.False = NapiRefHolder{this, GetBoolean(false)};
+  m_value.Global = NapiRefHolder{this, GetGlobal()};
+  m_value.Error = NapiRefHolder{this, GetProperty(m_value.Global, m_propertyId.Error)};
+}
+
+Value NapiJsiRuntime::evaluateJavaScript(const shared_ptr<const Buffer> &buffer, const string &sourceUrl) {
+  EnvScope envoScope{m_env};
+  napi_value script = CreateStringUtf8(buffer->data(), buffer->size());
+  napi_value result = RunScript(script, sourceUrl.c_str());
+
+  return ToJsiValue(result);
+}
+
+shared_ptr<const PreparedJavaScript> NapiJsiRuntime::prepareJavaScript(
+    const shared_ptr<const Buffer> &sourceBuffer,
+    string sourceUrl) {
+  EnvScope scope{m_env};
+  napi_value source = CreateStringUtf8(sourceBuffer->data(), sourceBuffer->size());
+  vector<uint8_t> serialized = SerializeScript(source, sourceUrl.c_str());
+  unique_ptr<const Buffer> serializedBuffer{new VectorBuffer{std::move(serialized)}};
+
+  return std::make_shared<NapiPreparedJavaScript>(std::move(serializedBuffer), sourceBuffer, std::move(sourceUrl));
```

Comment:
this seems wasteful (esp for large script buffers), can we do away with some of these str copies?

Reply:
Not sure. Are these copies necessary @tudorms, @vmoroz?

Reply:
@asklar, could you explain which string copy you see here are wasteful?

Reply:
1. `CreateStringUtf8` will call into napi which will copy the string
2. `SerializeScript` calls `vector::assign` with the 2nd overload, which copies the string

The VectorBuffer ctor does a move so that one is ok.

Reply:
Looks like this was merged with pending comments, @chrisglein can we update the bot to block merge until comments are resolved?


Filed #8650 to track the followup

Reply:
> Looks like this was merged with pending comments, @chrisglein can we update the bot to block merge until comments are resolved?

If there are sign offs and no blocking reviews, today it's up to the review author to decide whether all conversations need to be resolved. If a reviewer has something that should block the PR, then they can use the blocking state.

The bot we use doesn't have the ability to enforce conversation state. It'd have to be branch policy. [There is one for that](https://docs.github.com/en/repositories/configuring-branches-and-merges-in-your-repository/defining-the-mergeability-of-pull-requests/about-protected-branches#require-conversation-resolution-before-merging). But we'd want to ensure everyone was on board with that.

Reply:
@asklar, right, there is no extra copy: (1) creates the V8 value frm the provided buffer to access the script inside of V8, while the (2) returns the result by copying the compiled byte code from the V8 internal buffer into the result buffer.

Reply:
This comment missed my final sweep because it didn't show up in the files section after renaming the source file as `vnext/Microsoft.ReactNative.Cxx/JSI/NodeApiJsiRuntime.cpp`.
Will make sure to address all unresolved feedback in the future before setting `AutoMerge` :)
Note, this state of the code was already being used whenever we loaded V8, and is not a regression.

Thanks for creating the follow-up item. I'll assign to myself if it isn't already.

![image](https://user-images.githubusercontent.com/4507319/134060229-b5571a8b-36a5-44b3-acc3-fd6b1373a30f.png)